### PR TITLE
[ERE-2032] File extension whitelisting

### PR DESCRIPTION
**Pull request by ppettitau**

### CODE REVIEW

File: rdrf/rdrf/forms/dynamic/fields.py [hunk 1]
```diff
@@ -44,17 +45,23 @@ def validate(self, value):
...
             return super().validate(value)
-        __, ext = os.path.splitext(value._name)
+        __, ext = os.path.splitext(value.name)
         mime_type = magic.from_buffer(value.file.read(2048), mime=True)
         value.file.seek(0)
 
+        if not self._find_whitelisted_file_types(ext):
+            raise ValidationError(_(f"{ext} is not a supported file extension."))  # contact details if they believe it should be supported?
+
         blacklisted_mime_type = self._find_blacklisted_mime_type(mime_type)
```
Comment (ppettitau) [inline rdrf/rdrf/forms/dynamic/fields.py]: Do we want to get rid of the BlacklistedMimeType model and related logic?

Comment (ScriptSmith) [reply rdrf/rdrf/forms/dynamic/fields.py]: Yes, as per discussion

File: rdrf/rdrf/forms/dynamic/fields.py [hunk 2]
```diff
@@ -44,17 +45,23 @@ def validate(self, value):
...
     def validate(self, value):
         if not value:
             return super().validate(value)
-        __, ext = os.path.splitext(value._name)
+        __, ext = os.path.splitext(value.name)
         mime_type = magic.from_buffer(value.file.read(2048), mime=True)
         value.file.seek(0)
 
+        if not self._find_whitelisted_file_types(ext):
+            raise ValidationError(_(f"{ext} is not a supported file extension."))  # contact details if they believe it should be supported?
```
Comment (ScriptSmith) [inline rdrf/rdrf/forms/dynamic/fields.py]: Yes to the comment. Perhaps we need to add curator details as a field to the `Registry` object?

File: rdrf/rdrf/migrations/0159_whitelistedfileextension.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        WhitelistedFileExtension(file_extension='.msg'),
+        WhitelistedFileExtension(file_extension='.gz'),
+        WhitelistedFileExtension(file_extension='.zip'),
+        WhitelistedFileExtension(file_extension='.jfif'),
+        WhitelistedFileExtension(file_extension='.jpeg'),
+        WhitelistedFileExtension(file_extension='.jpg'),
+        WhitelistedFileExtension(file_extension='.tif'),
+        WhitelistedFileExtension(file_extension='.png'),
+        WhitelistedFileExtension(file_extension='.heic'),
+        WhitelistedFileExtension(file_extension='.mov')
```
Comment (ScriptSmith) [inline rdrf/rdrf/migrations/0159_whitelistedfileextension.py]: Some other media extensions found here:

https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types#common_image_file_types
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#browser_compatibility

```suggestion
        WhitelistedFileExtension(file_extension='.mov'),
        WhitelistedFileExtension(file_extension='.apng'),
        WhitelistedFileExtension(file_extension='.avif'),
        WhitelistedFileExtension(file_extension='.gif'),
        WhitelistedFileExtension(file_extension='.webp'),
        WhitelistedFileExtension(file_extension='.3gp'),
        WhitelistedFileExtension(file_extension='.aac'),
        WhitelistedFileExtension(file_extension='.flac'),
        WhitelistedFileExtension(file_extension='.mpg'),
        WhitelistedFileExtension(file_extension='.mpeg'),
        WhitelistedFileExtension(file_extension='.mp3'),
        WhitelistedFileExtension(file_extension='.mp4'),
        WhitelistedFileExtension(file_extension='.m4a'),
        WhitelistedFileExtension(file_extension='.m4p'),
        WhitelistedFileExtension(file_extension='.ogv'),
        WhitelistedFileExtension(file_extension='.ogg'),
        WhitelistedFileExtension(file_extension='.wav'),
        WhitelistedFileExtension(file_extension='.webm'),        
```

Could also allow powerpoint & excel files?

File: rdrf/rdrf/forms/dynamic/fields.py [hunk 4]
```diff
@@ -44,17 +45,23 @@ def validate(self, value):
...
+
     def _find_blacklisted_mime_type(self, mt):
         from rdrf.models.definition.models import BlacklistedMimeType
         return BlacklistedMimeType.objects.filter(mime_type=mt).first()
 
     def validate(self, value):
         if not value:
             return super().validate(value)
-        __, ext = os.path.splitext(value._name)
+        __, ext = os.path.splitext(value.name)
```
Comment (ScriptSmith) [inline rdrf/rdrf/forms/dynamic/fields.py:55]: Good catch. Not sure why it used the underscore value initially, might've just missed the `name` property.

**Review by ScriptSmith — CHANGES_REQUESTED** [review]

File: rdrf/rdrf/forms/dynamic/fields.py (hunk 2 above)
Comment (ppettitau) [reply rdrf/rdrf/forms/dynamic/fields.py]: Yes, I think that's a great idea.

Comment (ppettitau) [reply rdrf/rdrf/forms/dynamic/fields.py]: The file field `__init__` method is called when the application starts up, so I am not sure if I can pass in the Registry object to get the `contact_email` from it.

Any thoughts?

Comment (ScriptSmith) [reply rdrf/rdrf/forms/dynamic/fields.py]: We could say refer to the footer, and make sure we update that in all the registries.

File: rdrf/rdrf/forms/dynamic/fields.py (hunk 1 above)
Comment (ppettitau) [reply rdrf/rdrf/forms/dynamic/fields.py]: A new setting to capture the CURATOR_EMAIL has been created to support this. The registry repos will need to be updated with relevant contacts for mnd, arrk, angelman.

File: rdrf/rdrf/forms/dynamic/fields.py (hunk 2 above)
Comment (ppettitau) [reply rdrf/rdrf/forms/dynamic/fields.py]: A new setting to capture the CURATOR_EMAIL has been created to support this. The registry repos will need to be updated with relevant contacts for mnd, arrk, angelman.

File: rdrf/rdrf/forms/dynamic/fields.py (hunk 1 above)
Comment (ppettitau) [reply rdrf/rdrf/forms/dynamic/fields.py]: Blacklisted mime type logic has now been removed.

File: rdrf/rdrf/migrations/0159_whitelistedfileextension.py (hunk 3 above)
Comment (ppettitau) [reply rdrf/rdrf/migrations/0159_whitelistedfileextension.py]: Extra extensions now added.

**Review by ScriptSmith — APPROVED** [review]

**Pull request merged**